# Import relavant modules

In [16]:
import numpy as np
import pandas as pd
from fuzzywuzzy import fuzz
from fuzzywuzzy import process
import re
import warnings
warnings.filterwarnings("ignore")
import datamapper

In [2]:
pd.set_option('display.max_colwidth', -1)
pd.options.display.max_rows = 999

In [3]:
data = pd.read_csv('dataset.csv')

### Basic examination of data as df after a manual look

In [4]:
data.columns

Index(['id', 'name', 'tenant_id', 'address_line', 'city', 'state', 'zip',
       'country', 'phone', 'web', 'google_url', 'facebook_url', 'cars_url'],
      dtype='object')

In [651]:
data.head(5)

id                           name  tenant_id    address_line  \
0  3244033  # 1 Brookville Chevrolet       130469     1 E Main St      
1  18311    #1 BROOKVILLE CHEVROLET-BUICK  1840       ONE E MAIN ST    
2  118193   #1 BROOKVILLE CHEVROLET-BUICK  2980       1 East Main St   
3  18248    #1 BROOKVILLE CHEVROLET-BUICK  1836       ONE E MAIN ST    
4  575489   #1 BROOKVILLE CHEVROLET-BUICK  206        30 Main St       

         city state    zip country           phone  \
0  Brookville  PA    15825  US      (814) 849-8313   
1  BROOKVILLE  PA    15825  US      (814) 849-8313   
2  Brookville  PA    15825  US      (814) 849-8313   
3  BROOKVILLE  PA    15825  US      (814) 849-8313   
4  Brookville  PA    15825  US      (814) 849-8313   

                                  web  \
0  https://www.brookvillegmdealer.com   
1  NaN                                  
2  NaN                                  
3  NaN                                  
4  http://www.brookvillegmdealer.com    

                                              google_url  \
0  https://maps.google.com/maps?cid=12622059353105481409   
1  NaN                                                     
2  https://maps.google.com/maps?cid=12622059353105481409   
3  NaN                                                     
4  https://maps.google.com/maps?cid=12622059353105481409   

                                facebook_url  \
0  https://www.facebook.com/brookvillechevy/   
1  https://www.facebook.com/brookvillechevy/   
2  https://www.facebook.com/brookvillechevy/   
3  NaN                                         
4  https://www.facebook.com/brookvillechevy/   

                                                            cars_url  
0  https://www.cars.com/dealers/154863/1-brookville-chevrolet-buick/  
1  NaN                                                                
2  http://www.cars.com/dealers/154863/1-brookville-chevrolet-buick/   
3  NaN                                                                
4  https://www.cars.com/dealers/154863/1-brookville-chevrolet-buick/

In [652]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   id            30000 non-null  int64 
 1   name          30000 non-null  object
 2   tenant_id     30000 non-null  int64 
 3   address_line  29998 non-null  object
 4   city          30000 non-null  object
 5   state         30000 non-null  object
 6   zip           29999 non-null  object
 7   country       30000 non-null  object
 8   phone         28117 non-null  object
 9   web           17358 non-null  object
 10  google_url    26356 non-null  object
 11  facebook_url  19583 non-null  object
 12  cars_url      18962 non-null  object
dtypes: int64(2), object(11)
memory usage: 3.0+ MB


In [653]:
data.describe()

id      tenant_id
count  3.000000e+04  30000.000000 
mean   1.663171e+06  58770.750133 
std    1.119972e+06  50150.668922 
min    7.937000e+03  10.000000    
25%    5.751895e+05  4634.000000  
50%    1.521736e+06  45332.000000 
75%    2.777045e+06  115099.000000
max    3.454223e+06  139263.000000

In [654]:
data.shape

(30000, 13)

## Data cleaning
#### Standardize alphabet casing.
#### handle spaces and nulls for uniformity.
#### Convert single numbers from word to number.
#### Remove columns which is a same value through out DF.
#### Remove special charecters.
#### Handling nulls at important places.

In [656]:
data.columns

Index(['id', 'name', 'tenant_id', 'address_line', 'city', 'state', 'zip',
       'country', 'phone', 'web', 'google_url', 'facebook_url', 'cars_url'],
      dtype='object')

## Handling Non numeric columns

In [657]:
number_map = {
    'one': '1',
    'two': '2',
    'three': '3',
    'four': '4',
    'five': '5',
    'six': '6',
    'seven': '7',
    'eight': '8',
    'nine': '9',
    'zero' : '0'
}
number_map = {r"\b{}\b".format(k): v for k, v in number_map.items()}

In [658]:
non_numeric_columns = data.dtypes[data.dtypes != 'int64'][data.dtypes != 'float64'].index
non_numeric_columns

Index(['name', 'address_line', 'city', 'state', 'zip', 'country', 'phone',
       'web', 'google_url', 'facebook_url', 'cars_url'],
      dtype='object')

In [659]:
data[['name','address_line','zip','city','state','country']]=data[['name','address_line','zip','city','state','country']].replace('[^A-Za-z0-9]+',' ',regex=True)

for column in non_numeric_columns:
    data[column] = data[column].str.strip().str.lower()
    data[column] = data[column].replace(number_map, regex=True)
    data[column] = data[column].replace(r'^\s*$', np.NaN, regex=True)

    if (data[column].nunique() == 1) :
        data.drop([column], axis = 1, inplace = True)



Clearly as the Data is of dealers in U.S, the country column can be discarded

### Examining zip code uniformity

In [660]:
data[data['zip'].str.len().gt(5)]

## Incorrect zip, filling null

id                                  name  tenant_id  \
5      1575061  1 brookville chevrolet buick          54616       
7      1271984  1 brookville chevrolet buick          23144       
10     114365   1 cochran mazda                       1057        
25     1273660  1 cochran buick gmc of monroeville    23144       
40     1273111  1 cochran cadillac of monroeville     23144       
...        ...                                ...       ...       
29922  1146012  buena park nissan                     2552        
29945  1937104  buerkle honda                         10726       
29958  1275236  buff whelan chevrolet                 23144       
29961  94689    buffalo                               3848        
29988  1575294  buick gmc digital solution demo site  54616       

                address_line              city state         zip  \
5      30 main st             brookville        pa    15825 1642   
7      30 main st             brookville        pa    15825 1642   
10     4515 william penn hwy  monroeville       pa    15146 2813   
25     4520 william penn hwy  monroeville       pa    15146 2814   
40     4520 william penn hwy  monroeville       pa    15146 2814   
...                      ...          ...       ..           ...   
29922  6501 auto center dr    buena park        ca    90621 2901   
29945  3360 highway 61 n      saint paul        mn    55110 5212   
29958  40445 van dyke ave     sterling heights  mi    48313 3736   
29961  6755 transit rd        east amherst      ny    14051 1485   
29988  2200 1st ave s         seattle           wa    98134 1452   

                phone                                  web  \
5      (814) 246-2055  http://www.brookvillegmdealer.com     
7      (877) 223-2395  http://www.brookvillegmdealer.com     
10     (412) 373-3333  NaN                                   
25     (877) 304-9607  http://www.cochranofmonroeville.com   
40     (877) 824-2498  http://www.cochrancadillac.com        
...               ...                             ...        
29922  (714) 739-0800  NaN                                   
29945  (651) 490-6600  https://www.buerklehonda.com/         
29958  (586) 933-5557  http://www.buffwhelan.com             
29961  (716) 636-6000  NaN                                   
29988  (206) 263-6969  NaN                                   

                                                  google_url  \
5      https://maps.google.com/maps?cid=12622059353105481409   
7      https://maps.google.com/maps?cid=12622059353105481409   
10     https://maps.google.com/maps?cid=12743786671890848048   
25     https://maps.google.com/maps?cid=2752462366417878165    
40     https://maps.google.com/maps?cid=17115351350047781923   
...                                                      ...   
29922  https://maps.google.com/maps?cid=12617323471123278991   
29945  https://maps.google.com/maps?cid=885141847253462911     
29958  https://maps.google.com/maps?cid=8836726731726321501    
29961  https://maps.google.com/maps?cid=9746768870518867160    
29988  NaN                                                     

                                        facebook_url  \
5      https://www.facebook.com/brookvillechevy/       
7      https://www.facebook.com/brookvillechevy/       
10     https://www.facebook.com/numberonecochran/      
25     NaN                                             
40     NaN                                             
...    ...                                             
29922  https://www.facebook.com/buenaparknissan/       
29945  https://www.facebook.com/buerklehondawbl/       
29958  https://www.facebook.com/nickcarriercarsales/   
29961  https://www.facebook.com/195215093949157        
29988  NaN                                             

                                                                   cars_url  
5      https://www.cars.com/dealers/154863/1-brookville-chevrolet-buick/     
7      https://www.cars.com/dealers/154863/1-br

In [661]:
## Lets derive a short zip from long zips
data['zip_short']  = data.zip.apply(lambda x: str(x).split(' ')[0] if len(str(x))>5 else x)

In [662]:
data.zip = data.zip.apply(lambda x: x if str(x).replace(' ','').isdigit() else np.nan)
data.zip_short = data.zip_short.apply(lambda x: x if str(x).replace(' ','').isdigit() else np.nan)

In [663]:
data.zip.isnull().sum()

2

In [664]:
data.zip_short.isnull().sum()

2

Seems to be a bad row, but the zip and address line can be extracted using Geocoding API from google maps url, which is too advanced for now, so removing it instead

In [665]:
data = data[~data.zip.isnull()]

In [666]:
data.isnull().sum()

id              0    
name            2    
tenant_id       0    
address_line    1    
city            0    
state           0    
zip             0    
phone           1883 
web             12642
google_url      3644 
facebook_url    10415
cars_url        11036
zip_short       0    
dtype: int64

# Generalize web addresses only upto .com

In [667]:
data.web.head(10)

0    https://www.brookvillegmdealer.com
1    NaN                               
2    NaN                               
3    NaN                               
4    http://www.brookvillegmdealer.com 
5    http://www.brookvillegmdealer.com 
6    NaN                               
7    http://www.brookvillegmdealer.com 
8    https://www.brookvillegmdealer.com
9    http://www.ricartkia.com/         
Name: web, dtype: object

In [668]:
data.web = data.web.str.strip().str.replace('https://','').str.replace('http://','')
data.web

0        www.brookvillegmdealer.com
1        NaN                       
2        NaN                       
3        NaN                       
4        www.brookvillegmdealer.com
                    ...            
29995    NaN                       
29996    www.buickgmcfalls.com/    
29997    NaN                       
29998    www.buickgmcfalls.com     
29999    NaN                       
Name: web, Length: 29998, dtype: object

In [669]:
data.web = data.web.apply(lambda x: str(x).split('/')[0] if x is not None else np.nan)
data.web = data.web.replace('nan',np.nan)

In [670]:
data.web.head(10)

0    www.brookvillegmdealer.com
1    NaN                       
2    NaN                       
3    NaN                       
4    www.brookvillegmdealer.com
5    www.brookvillegmdealer.com
6    NaN                       
7    www.brookvillegmdealer.com
8    www.brookvillegmdealer.com
9    www.ricartkia.com         
Name: web, dtype: object

In [671]:
data.web.isnull().sum()

12642

## Data with no dealer name available

In [672]:
data[data.name.isnull()] 

# These rows are useless as well. removing them.

id name  tenant_id address_line        city state    zip phone  web  \
160  2728616  NaN  109662     755 i 45     huntsville  tx    77340  NaN   NaN   
161  2733888  NaN  114950     755 i 45     huntsville  tx    77340  NaN   NaN   

                                               google_url facebook_url  \
160  NaN                                                   NaN           
161  https://maps.google.com/maps?cid=7613114450716243350  NaN           

    cars_url zip_short  
160  NaN      77340     
161  NaN      77340

In [673]:
data = data[~data.name.isnull()]

In [674]:
data.isnull().sum()

id              0    
name            0    
tenant_id       0    
address_line    1    
city            0    
state           0    
zip             0    
phone           1881 
web             12640
google_url      3643 
facebook_url    10413
cars_url        11034
zip_short       0    
dtype: int64

In [675]:
data[data.address_line.isnull()]

# Except addressline, everything is present, lets retain this for further uses

id                name  tenant_id address_line  \
1707  2515038  acura turnersville  109662     NaN           

                     city state   zip           phone  \
1707  washington township  nj    8012  (856) 497-4749   

                            web  \
1707  www.acuraturnersville.com   

                                                google_url facebook_url  \
1707  https://maps.google.com/maps?cid=2201156893789197068  NaN           

     cars_url zip_short  
1707  NaN      8012

In [676]:
data.shape

(29996, 13)

### Similarly phone number can be standardized if there are any issues. Leaving that for now

# Inspect no of unique entries per columns

In [677]:
for column in data.columns:
    print("The " + column  + " has following unique elements :" + str(data[column].nunique()) + " out of " + str(data.shape[0]) + " entries")

The id has following unique elements :29664 out of 29996 entries
The name has following unique elements :7970 out of 29996 entries
The tenant_id has following unique elements :661 out of 29996 entries
The address_line has following unique elements :7916 out of 29996 entries
The city has following unique elements :1882 out of 29996 entries
The state has following unique elements :53 out of 29996 entries
The zip has following unique elements :4264 out of 29996 entries
The phone has following unique elements :7657 out of 29996 entries
The web has following unique elements :4770 out of 29996 entries
The google_url has following unique elements :5546 out of 29996 entries
The facebook_url has following unique elements :7878 out of 29996 entries
The cars_url has following unique elements :4986 out of 29996 entries
The zip_short has following unique elements :2750 out of 29996 entries


### As first step remove duplicates compared in whole data

In [678]:
duplicate_whole = data[data.duplicated(data.columns)]

In [679]:
duplicate_whole.shape

(32, 13)

In [680]:
data = data.drop_duplicates(subset = data.columns, keep = 'first')
data.shape

(29964, 13)

In [681]:
data.isnull().sum()

id              0    
name            0    
tenant_id       0    
address_line    1    
city            0    
state           0    
zip             0    
phone           1881 
web             12640
google_url      3643 
facebook_url    10413
cars_url        11034
zip_short       0    
dtype: int64

### There seems to be that there are few duplicates based on ID, lets inspect them if they are pure duplicates or is there any info we can fetch from them?

In [682]:
duplicate = data[data['id'].duplicated(keep=False)]
duplicate.shape

(509, 13)

###  Upon manual inspection in file it can be seen these are pure duplicates except for changes in URLs, lets double check using the other metadata except URLs

In [683]:
duplicate2 = data[data.duplicated(subset = ['id','name','tenant_id','address_line','city','state','zip','phone'], keep=False)]
duplicate2.shape

(509, 13)

Our assumption turned out to be true, there are 509 duplicates where the only variations lie at URLs like web google_url etc. As we dont exactly know which URL is correct, lets pick one value among them and lets merge tenants.

In [684]:
duplicate2.isnull().sum()

id              0  
name            0  
tenant_id       0  
address_line    0  
city            0  
state           0  
zip             0  
phone           24 
web             269
google_url      22 
facebook_url    115
cars_url        53 
zip_short       0  
dtype: int64

## Try to fill information from similar rows at nulls, which seem to had no impact over nulls.

# Go ahead and remove duplicates based on metadata and retain only 1 value

In [685]:
data = data.drop_duplicates(subset = ['id','name','tenant_id','address_line','city','state','zip','phone'], keep = 'first')
data.shape

(29664, 13)

We have a reduce of over 350 entries.

Lets check number of dealers based on name

In [686]:
data.isnull().sum()

id              0    
name            0    
tenant_id       0    
address_line    1    
city            0    
state           0    
zip             0    
phone           1869 
web             12495
google_url      3632 
facebook_url    10352
cars_url        11004
zip_short       0    
dtype: int64

#### inspection and consolidation based on name

In [687]:
data.name.value_counts()

bruckner s mack volvo                     35
affordable auto sales                     29
affordable autos                          28
anderson ford                             27
all american chevrolet                    25
                                          ..
bayway cadillac of the woodlands 17010    1 
b j maurer motor company 07428            1 
bosak motor sales                         1 
bergstrom enterprise motorcars            1 
audi of roseville                         1 
Name: name, Length: 7970, dtype: int64

In [688]:
data[data.name == "bruckner s mack volvo"]

id                   name  tenant_id            address_line  \
29402  1605023  bruckner s mack volvo  86817      1125 slaton rd           
29403  1594486  bruckner s mack volvo  57705      8101 daytona rd nw       
29404  2739739  bruckner s mack volvo  114950     4050 bloomfield hwy      
29405  1507833  bruckner s mack volvo  45332      4275 meadow ln           
29406  1594623  bruckner s mack volvo  57705      4050 bloomfield hwy      
29407  2743175  bruckner s mack volvo  115099     2509 n taylor ave        
29408  1512170  bruckner s mack volvo  45332      2121 nw county rd        
29409  1518181  bruckner s mack volvo  45332      9471 i 40 east exit 76   
29410  1603756  bruckner s mack volvo  86817      2509 n taylor ave        
29411  2735232  bruckner s mack volvo  114950     2121 nw county rd        
29412  2748099  bruckner s mack volvo  115107     2441 e interstate 20     
29413  2799261  bruckner s mack volvo  115107     9471 i 40 east exit 76   
29414  1523637  bruckner s mack volvo  45332      829 n bell st            
29415  1515290  bruckner s mack volvo  45332      2441 e interstate 20     
29416  1604787  bruckner s mack volvo  86817      2602 millhaven rd        
29417  1521640  bruckner s mack volvo  45332      1125 slaton rd           
29418  1509015  bruckner s mack volvo  45332      4050 bloomfield hwy      
29419  1597180  bruckner s mack volvo  57705      2441 e interstate 20     
29420  2741320  bruckner s mack volvo  115099     1125 slaton rd           
29421  2741729  bruckner s mack volvo  115099     829 n bell st            
29422  2747092  bruckner s mack volvo  115099     9471 i 40 east exit 76   
29423  1512025  bruckner s mack volvo  45332      2509 n taylor ave        
29424  1521134  bruckner s mack volvo  45332      2602 millhaven rd        
29425  2741207  bruckner s mack volvo  115099     2602 millhaven rd        
29426  1513651  bruckner s mack volvo  45332      2101 commerce pkwy       
29427  1593747  bruckner s mack volvo  57705      4275 meadow ln           
29428  1605639  bruckner s mack volvo  86817      829 n bell st            
29429  2739541  bruckner s mack volvo  114950     8101 daytona rd nw       
29430  1600156  bruckner s mack volvo  83249      9471 i 40 east exit 76   
29431  2751363  bruckner s mack volvo  115107     2101 commerce pkwy       
29432  1508929  bruckner s mack volvo  45332      8101 daytona rd nw       
29433  2738849  bruckner s mack volvo  114950     4275 meadow ln           
29434  2731323  bruckner s mack volvo  109662     2121 nw county rd        
29435  1591568  bruckner s mack volvo  54617      2121 nw county rd        
29436  1596595  bruckner s mack volvo  57705      2101 commerce pkwy       

               city state    zip           phone                    web  \
29402  lubbock       tx    79404  (888) 999-0491  www.brucknertruck.com   
29403  albuquerque   nm    87121  (877) 999-2451  www.brucknertruck.com   
29404  farmington    nm    87401  (877) 999-2971  NaN                     
29405  bossier city  la    71111  (866) 928-6580  www.brucknertruck.com   
29406  farmington    nm    87401  (877) 999-2971  www.brucknertruck.com   
29407  garden city   ks    67846  (888) 999-7810  NaN                     
29408  hobbs         nm    88240  (877) 999-1440  www.brucknertruck.com   
29409  amarillo      tx    79118  (877) 999-8071  www.brucknertruck.com   
29410  garden city   ks    67846  (888) 999-7810  www.brucknertruck.com   
29411  hobbs         nm    88240  (877) 999-1440  NaN                     
29412  odessa        tx    79766  (866) 342-2462  NaN                     
29413  amarillo      tx    79118  (877) 999-8071  NaN                     
29414  san angelo    tx    76903  (877) 841-8261  www.brucknertruck.com   
29415  odessa        tx    79766  (866) 342-2462  www.brucknertruck.com   
29416  monroe        la    71203  (866) 897-8274  www.brucknertruck.com   
29417  lubbock       tx    79404  (888) 999-0491  www.brucknertruck.com   


# There seem to be having entries with same name, address, city and state but different IDs, such entries can be consolidated.

In [689]:
data['meta_data'] = data['name'].fillna(' ') + ',' + data['address_line'].fillna(' ') + ',' + data['city'].fillna(' ')\
+data['state'].fillna(' ') + ',' + data['zip_short'].fillna(' ')

In [690]:
data['meta_data']

0        1 brookville chevrolet,1 e main st,brookvillepa,15825                  
1        1 brookville chevrolet buick,1 e main st,brookvillepa,15825            
2        1 brookville chevrolet buick,1 east main st,brookvillepa,15825         
3        1 brookville chevrolet buick,1 e main st,brookvillepa,15825            
4        1 brookville chevrolet buick,30 main st,brookvillepa,15825             
                                    ...                                         
29995    buick gmc menomonee falls,n88 w14167 main st,menomonee fallswi,53051   
29996    buick gmc menomonee falls,n88 w14167 main st,menomonee fallswi,53051   
29997    buick gmc menomonee falls,main st,menomonee fallswi,53051              
29998    buick gmc menomonee falls,n88 w14167 main st,menomonee fallswi,53051   
29999    buick gmc menomonee falls,n88 w14167 main st,n88 w14167 main stwi,53051
Name: meta_data, Length: 29664, dtype: object

lets also map numbers from 1 to 9 with corresponding value

In [691]:
data["meta_data"].nunique()

13523

### Out of 29664 entries only 13523 entries are unique which means around 50% of data is repeated based on only name,address,city and state. Lets work on them

In [692]:
duplicated_metadata =  data[data.duplicated(subset = ['meta_data'], keep=False)]

In [693]:
duplicated_metadata.shape

(21920, 14)

In [694]:
duplicated_metadata

id                          name  tenant_id  \
1      18311    1 brookville chevrolet buick  1840        
2      118193   1 brookville chevrolet buick  2980        
3      18248    1 brookville chevrolet buick  1836        
4      575489   1 brookville chevrolet buick  206         
5      1575061  1 brookville chevrolet buick  54616       
...        ...                           ...    ...       
29992  1519947  buick gmc fort walton beach   45332       
29993  3250701  buick gmc fort walton beach   130469      
29995  2747944  buick gmc menomonee falls     115107      
29996  1596983  buick gmc menomonee falls     57705       
29998  3234870  buick gmc menomonee falls     130469      

                    address_line               city state         zip  \
1      1 e main st                brookville         pa    15825        
2      1 east main st             brookville         pa    15825        
3      1 e main st                brookville         pa    15825        
4      30 main st                 brookville         pa    15825        
5      30 main st                 brookville         pa    15825 1642   
...           ...                        ...         ..           ...   
29992  329 miracle strip pkwy sw  fort walton beach  fl    32548        
29993  329 miracle strip pkwy sw  fort walton beach  fl    32548        
29995  n88 w14167 main st         menomonee falls    wi    53051        
29996  n88 w14167 main st         menomonee falls    wi    53051        
29998  n88 w14167 main st         menomonee falls    wi    53051        

                phone                              web  \
1      (814) 849-8313  NaN                               
2      (814) 849-8313  NaN                               
3      (814) 849-8313  NaN                               
4      (814) 849-8313  www.brookvillegmdealer.com        
5      (814) 246-2055  www.brookvillegmdealer.com        
...               ...                         ...        
29992  (850) 244-5165  www.buickgmcfortwaltonbeach.com   
29993  (850) 364-1032  www.buickgmcfortwaltonbeach.com   
29995  (262) 255-6000  NaN                               
29996  (262) 255-6000  www.buickgmcfalls.com             
29998  (262) 255-6000  www.buickgmcfalls.com             

                                                  google_url  \
1      NaN                                                     
2      https://maps.google.com/maps?cid=12622059353105481409   
3      NaN                                                     
4      https://maps.google.com/maps?cid=12622059353105481409   
5      https://maps.google.com/maps?cid=12622059353105481409   
...                                                      ...   
29992  https://maps.google.com/maps?cid=1201605996612827940    
29993  https://maps.google.com/maps?cid=4601099509292193902    
29995  https://maps.google.com/maps?cid=10136321579153410390   
29996  https://maps.google.com/maps?cid=10136321579153410390   
29998  https://maps.google.com/maps?cid=10136321579153410390   

                                                             facebook_url  \
1      https://www.facebook.com/brookvillechevy/                            
2      https://www.facebook.com/brookvillechevy/                            
3      NaN                                                                  
4      https://www.facebook.com/brookvillechevy/                            
5      https://www.facebook.com/brookvillechevy/                            
...                                          ...                            
29992  https://www.facebook.com/buickcadillacgmcftwalton                    
29993  https://www.facebook.com/buickcadillacgmcftwalton/                   
29995  https://www.facebook.com/buick-gmc-menomonee-falls-276258956513411   
29996  https://www.facebook.com/ernievonschledornbuickgmc/                  
29998  https://www.facebook.com/ernievonschledornbuickgmc/                  

                                   

In [695]:
data_combined=data.groupby(['meta_data']).first().reset_index()

In [696]:
data_combined.shape

(13523, 14)

In [697]:
data_combined

meta_data  \
0      001 putnam chevrolet and cadillac,3052 martin luther king junior blvd,burlingameca,94031   
1      002 fremont chevrolet,5850 cushing pkwy,fremontca,94538                                    
2      003 chevrolet of stevens creek,3640 stevens creek blvd,san joseca,95117                    
3      004 capitol chevrolet,905 capitol expressway auto mall,san joseca,95136                    
4      1 brookville chevrolet buick,1 e main st,brookvillepa,15825                                
...                                                            ...                                
13518  byers kia oh057,5342 columbus pike,lewis centeroh,43035                                    
13519  byers mazda,2455 billingsley rd,columbusoh,43235                                           
13520  byford buick gmc llc,2900 s hwy 81,chickashaok,73018                                       
13521  byford buick gmc llc,2900 u s 81,chickashaok,73018                                         
13522  byford chevrolet buick gmc,1909 u s 81,duncanok,73533                                      

            id                               name  tenant_id  \
0      1123420  001 putnam chevrolet and cadillac  11138       
1      1123421  002 fremont chevrolet              11138       
2      1123422  003 chevrolet of stevens creek     11138       
3      1123423  004 capitol chevrolet              11138       
4      18311    1 brookville chevrolet buick       1840        
...      ...                             ...        ...        
13518  16039    byers kia oh057                    1157        
13519  208328   byers mazda                        1057        
13520  13710    byford buick gmc llc               1836        
13521  576069   byford buick gmc llc               206         
13522  576061   byford chevrolet buick gmc         206         

                              address_line          city state         zip  \
0      3052 martin luther king junior blvd  burlingame    ca    94031        
1      5850 cushing pkwy                    fremont       ca    94538        
2      3640 stevens creek blvd              san jose      ca    95117        
3      905 capitol expressway auto mall     san jose      ca    95136        
4      1 e main st                          brookville    pa    15825        
...            ...                                 ...    ..      ...        
13518  5342 columbus pike                   lewis center  oh    43035        
13519  2455 billingsley rd                  columbus      oh    43235 1925   
13520  2900 s hwy 81                        chickasha     ok    73018        
13521  2900 u s 81                          chickasha     ok    73018        
13522  1909 u s 81                          duncan        ok    73533        

                phone                 web  \
0      (479) 856-6390  www.bigotires.com    
1      (928) 344-2702  NaN                  
2      (520) 625-9414  NaN                  
3      (928) 776-1111  NaN                  
4      (814) 849-8313  NaN                  
...               ...  ...                  
13518  NaN             www.byerskia.com     
13519  (614) 792-2455  NaN                  
13520  (405) 224-0475  NaN                  
13521  (405) 224-0475  www.byfordauto.com   
13522  (580) 255-0550  www.byfordgm.com     

                                                  google_url  \
0      NaN                                                     
1      https://maps.google.com/maps?cid=2529089401223706257    
2      https://maps.google.com/maps?cid=17294707326087392923   
3      https://maps.google.com/maps?cid=16621265989335969999   
4      NaN                                                     
...    ...                                                     
13518  https://maps.google.com/maps?cid=16395390017796853957   
13519  https://maps.google.com/maps?cid=1129683514267478914    
13520  NaN                                                     
13521  NaN

In [698]:
data_combined.shape

(13523, 14)

### The number of distinct dealers is now down to 13523. which is 51% of actual data. Lets check if we can refine further

In [699]:
data_combined

meta_data  \
0      001 putnam chevrolet and cadillac,3052 martin luther king junior blvd,burlingameca,94031   
1      002 fremont chevrolet,5850 cushing pkwy,fremontca,94538                                    
2      003 chevrolet of stevens creek,3640 stevens creek blvd,san joseca,95117                    
3      004 capitol chevrolet,905 capitol expressway auto mall,san joseca,95136                    
4      1 brookville chevrolet buick,1 e main st,brookvillepa,15825                                
...                                                            ...                                
13518  byers kia oh057,5342 columbus pike,lewis centeroh,43035                                    
13519  byers mazda,2455 billingsley rd,columbusoh,43235                                           
13520  byford buick gmc llc,2900 s hwy 81,chickashaok,73018                                       
13521  byford buick gmc llc,2900 u s 81,chickashaok,73018                                         
13522  byford chevrolet buick gmc,1909 u s 81,duncanok,73533                                      

            id                               name  tenant_id  \
0      1123420  001 putnam chevrolet and cadillac  11138       
1      1123421  002 fremont chevrolet              11138       
2      1123422  003 chevrolet of stevens creek     11138       
3      1123423  004 capitol chevrolet              11138       
4      18311    1 brookville chevrolet buick       1840        
...      ...                             ...        ...        
13518  16039    byers kia oh057                    1157        
13519  208328   byers mazda                        1057        
13520  13710    byford buick gmc llc               1836        
13521  576069   byford buick gmc llc               206         
13522  576061   byford chevrolet buick gmc         206         

                              address_line          city state         zip  \
0      3052 martin luther king junior blvd  burlingame    ca    94031        
1      5850 cushing pkwy                    fremont       ca    94538        
2      3640 stevens creek blvd              san jose      ca    95117        
3      905 capitol expressway auto mall     san jose      ca    95136        
4      1 e main st                          brookville    pa    15825        
...            ...                                 ...    ..      ...        
13518  5342 columbus pike                   lewis center  oh    43035        
13519  2455 billingsley rd                  columbus      oh    43235 1925   
13520  2900 s hwy 81                        chickasha     ok    73018        
13521  2900 u s 81                          chickasha     ok    73018        
13522  1909 u s 81                          duncan        ok    73533        

                phone                 web  \
0      (479) 856-6390  www.bigotires.com    
1      (928) 344-2702  NaN                  
2      (520) 625-9414  NaN                  
3      (928) 776-1111  NaN                  
4      (814) 849-8313  NaN                  
...               ...  ...                  
13518  NaN             www.byerskia.com     
13519  (614) 792-2455  NaN                  
13520  (405) 224-0475  NaN                  
13521  (405) 224-0475  www.byfordauto.com   
13522  (580) 255-0550  www.byfordgm.com     

                                                  google_url  \
0      NaN                                                     
1      https://maps.google.com/maps?cid=2529089401223706257    
2      https://maps.google.com/maps?cid=17294707326087392923   
3      https://maps.google.com/maps?cid=16621265989335969999   
4      NaN                                                     
...    ...                                                     
13518  https://maps.google.com/maps?cid=16395390017796853957   
13519  https://maps.google.com/maps?cid=1129683514267478914    
13520  NaN                                                     
13521  NaN

In [700]:
data_combined.isnull().sum()

meta_data       0   
id              0   
name            0   
tenant_id       0   
address_line    1   
city            0   
state           0   
zip             0   
phone           1141
web             5057
google_url      1711
facebook_url    5634
cars_url        5867
zip_short       0   
dtype: int64

In [701]:
data_combined.count()

meta_data       13523
id              13523
name            13523
tenant_id       13523
address_line    13522
city            13523
state           13523
zip             13523
phone           12382
web             8466 
google_url      11812
facebook_url    7889 
cars_url        7656 
zip_short       13523
dtype: int64

In [702]:
for column in data_combined.columns:
    print("The " + column  + " has following unique elements :" + str(data_combined[column].nunique()) + " out of " + str(data_combined.shape[0]) + " entries")

The meta_data has following unique elements :13523 out of 13523 entries
The id has following unique elements :13523 out of 13523 entries
The name has following unique elements :7970 out of 13523 entries
The tenant_id has following unique elements :481 out of 13523 entries
The address_line has following unique elements :7916 out of 13523 entries
The city has following unique elements :1882 out of 13523 entries
The state has following unique elements :53 out of 13523 entries
The zip has following unique elements :3635 out of 13523 entries
The phone has following unique elements :6287 out of 13523 entries
The web has following unique elements :4450 out of 13523 entries
The google_url has following unique elements :5249 out of 13523 entries
The facebook_url has following unique elements :5651 out of 13523 entries
The cars_url has following unique elements :4388 out of 13523 entries
The zip_short has following unique elements :2750 out of 13523 entries


## Name and address uniquely represent a specific dealer, lets handle overlapping among them.

In [703]:
data_combined['dealer_addresses'] = data_combined['name'] + ',' + data_combined['address_line']

In [704]:
data_combined['dealer_addresses'].value_counts()

audi turnersville,3751 nj 42                                   4
bass mazda,4900 detroit rd                                     4
audi pembroke pines,15000 sheridan st                          4
bell ford,1200 nj 27                                           4
a crivelli buick gmc,1403 allegheny blvd                       4
                                                              ..
bergstrom chevrolet of middleton,3605 tribeca dr               1
autonation chevrolet north richland hills,7769 boulevard 26    1
brotherton cadillac nw,17545 aurora avenue north               1
audi la crosse,726 monitor st                                  1
brown brown chevrolet,145 east main street                     1
Name: dealer_addresses, Length: 11395, dtype: int64

In [705]:
data_combined_v2=data_combined.groupby(['dealer_addresses']).first().reset_index()

In [706]:
data_combined_v2.shape

(11395, 15)

### 3000 entries further reduced.

In [707]:
data_combined_v2.isnull().sum()

dealer_addresses    0   
meta_data           0   
id                  0   
name                0   
tenant_id           0   
address_line        0   
city                0   
state               0   
zip                 0   
phone               1115
web                 3178
google_url          1681
facebook_url        3761
cars_url            4012
zip_short           0   
dtype: int64

In [708]:
for column in data_combined_v2.columns:
    print("The " + column  + " has following unique elements :" + str(data_combined_v2[column].nunique()) + " out of " + str(data_combined_v2.shape[0]) + " entries")

The dealer_addresses has following unique elements :11395 out of 11395 entries
The meta_data has following unique elements :11395 out of 11395 entries
The id has following unique elements :11395 out of 11395 entries
The name has following unique elements :7970 out of 11395 entries
The tenant_id has following unique elements :468 out of 11395 entries
The address_line has following unique elements :7916 out of 11395 entries
The city has following unique elements :1842 out of 11395 entries
The state has following unique elements :53 out of 11395 entries
The zip has following unique elements :3491 out of 11395 entries
The phone has following unique elements :6032 out of 11395 entries
The web has following unique elements :4424 out of 11395 entries
The google_url has following unique elements :5197 out of 11395 entries
The facebook_url has following unique elements :5565 out of 11395 entries
The cars_url has following unique elements :4336 out of 11395 entries
The zip_short has following un

## The uniqueness in phone number is still quite low,

### Single phone number cannot be used by multiple dealers, but there is still scope that a single person is running two dealerships, so having only phone as primary key will not make sense, lets try to combine it with name,addressline and try to club it together to make it a unique entity

In [709]:
data_combined_v2['dealer_contact'] = data_combined_v2['name'].fillna(' ') + ',' + data_combined_v2['address_line'].fillna(' ') +  ',' + data_combined_v2['phone'].fillna(' ')

In [710]:
data_combined_v2['dealer_contact'].nunique()

11395

In [711]:
data_combined_v2['dealer_contact'].value_counts().sort_index(ascending=False)


byford chevrolet buick gmc,1909 u s 81,(580) 255-0550                                   1
byford buick gmc llc,2900 u s 81,(405) 224-0475                                         1
byford buick gmc llc,2900 s hwy 81,(405) 224-0475                                       1
byers mazda,2455 billingsley rd,(614) 792-2455                                          1
byers kia oh057,5342 columbus pike,                                                     1
                                                                                       ..
1 brookville chevrolet buick,1 e main st,(814) 849-8313                                 1
004 capitol chevrolet,905 capitol expressway auto mall,(928) 776-1111                   1
003 chevrolet of stevens creek,3640 stevens creek blvd,(520) 625-9414                   1
002 fremont chevrolet,5850 cushing pkwy,(928) 344-2702                                  1
001 putnam chevrolet and cadillac,3052 martin luther king junior blvd,(479) 856-6390    1
Name: deal

In [712]:
data_combined_v2.shape

(11395, 16)

##### Not much of a result after refining here

## It is clearly visible that still grouping can be done to the data as the unique entries in few columns which are important to identify unique dealer are quite less, for example,combination of website,google and facebook. Lets try to examine that

In [713]:
data_combined_v2.web.value_counts()

www.billionauto.com                       23
www.brucknertruck.com                     13
www.autonationcollisioncenters.com        11
www.autonationcadillacwestamarillo.com    8 
www.honda.com                             8 
                                         .. 
www.acgiant.com                           1 
www.bobmaxeylincoln.com                   1 
www.cochrannissanmonroeville.com          1 
www.bommaritonissanwest.com               1 
www.audimv.com                            1 
Name: web, Length: 4424, dtype: int64

In [635]:
data_combined_v2[data_combined_v2.web=='http://www.billionauto.com/']

Empty DataFrame
Columns: [dealer_addresses, meta_data, id, name, tenant_id, address_line, city, state, zip, phone, web, google_url, facebook_url, cars_url, zip_short, dealer_contact]
Index: []

In [714]:
duplicated_internet_presence =  data_combined_v2[data_combined_v2.duplicated(subset = ['web','google_url','facebook_url','cars_url'], keep=False)]

In [715]:
duplicated_internet_presence.shape

(2482, 16)

In [716]:
duplicated_internet_presence

dealer_addresses  \
6      1 brookville chevrolet buick,30 main st          
7      1 brookville chevrolet,1 e main st               
8      1 cochran 5,5200 campbells run rd                
15     1 cochran ford of allegheny valley,110 rte 908   
17     1 cochran ford,110 rte 908 building 1            
...                                      ...            
11360  burns chevrolet inc,1733 north limestone st      
11361  burnsville chevrolet buick inc,627 w main st     
11363  burnsville chevrolet buick,627 w main st         
11375  buss automotive inc,1255 e green bay st          
11392  byford buick gmc llc,2900 s hwy 81               

                                                                    meta_data  \
6      1 brookville chevrolet buick,30 main st,brookvillepa,15825               
7      1 brookville chevrolet,1 e main st,brookvillepa,15825                    
8      1 cochran 5,5200 campbells run rd,pittsburghpa,15205                     
15     1 cochran ford of allegheny valley,110 rte 908,natrona heightsca,15065   
17     1 cochran ford,110 rte 908 building 1,natrona heightspa,15065            
...                                                              ...            
11360  burns chevrolet inc,1733 north limestone st,gaffneysc,29340              
11361  burnsville chevrolet buick inc,627 w main st,burnsvillenc,28714          
11363  burnsville chevrolet buick,627 w main st,burnsvillenc,28714              
11375  buss automotive inc,1255 e green bay st,shawanowi,54166                  
11392  byford buick gmc llc,2900 s hwy 81,chickashaok,73018                     

            id                                name  tenant_id  \
6      575489   1 brookville chevrolet buick        206         
7      3244033  1 brookville chevrolet              130469      
8      436256   1 cochran 5                         8552        
15     1488522  1 cochran ford of allegheny valley  44849       
17     3233063  1 cochran ford                      130469      
...        ...             ...                         ...      
11360  45079    burns chevrolet inc                 2980        
11361  18220    burnsville chevrolet buick inc      1836        
11363  97092    burnsville chevrolet buick          1840        
11375  24884    buss automotive inc                 1836        
11392  13710    byford buick gmc llc                1836        

                  address_line             city state         zip  \
6      30 main st               brookville       pa    15825        
7      1 e main st              brookville       pa    15825        
8      5200 campbells run rd    pittsburgh       pa    15205        
15     110 rte 908              natrona heights  ca    15065        
17     110 rte 908 building 1   natrona heights  pa    15065 2810   
...                       ...               ...  ..           ...   
11360  1733 north limestone st  gaffney          sc    29340        
11361  627 w main st            burnsville       nc    28714        
11363  627 w main st            burnsville       nc    28714        
11375  1255 e green bay st      shawano          wi    54166        
11392  2900 s hwy 81            chickasha        ok    73018        

                phone                                web  \
6      (814) 849-8313  www.brookvillegmdealer.com          
7      (814) 849-8313  www.brookvillegmdealer.com          
8      (877) 290-1514  NaN                                 
15     (877) 844-2719  www.cochranfordnatronaheights.com   
17     (412) 245-4320  www.cochranfordnatronaheights.com   
...               ...                                ...   
11360  (864) 489-6000  NaN                                 
11361  (828) 682-6141  NaN                                 
11363  NaN             NaN                                 
11375  (715) 524-0444  NaN                                 
11392  (405) 224-0475  NaN                                 

                                                

## Clearly visible that few of them belong to same dealer, lets work out on consolidating them.

In [717]:
data_combined_v2['internet_presence'] = data['name'].fillna(' ') + ',' + data['phone'].fillna(' ') + ',' + data['zip_short'].fillna(' ') + ',' + data_combined_v2['web'].fillna(' ') + ',' + data_combined_v2['google_url'].fillna(' ') + ',' + data_combined_v2['facebook_url'].fillna(' ')\
+data_combined_v2['cars_url'].fillna(' ')

In [718]:
data_combined_v2['internet_presence'].value_counts()

auffenberg chevrolet buick gmc,(573) 747-4120,63640, , ,                                                                                                                                                                                                            4
autonation ford westlake 2025, ,44145, , ,                                                                                                                                                                                                                          3
autonation ford westlake,(440) 296-3019,44145, , ,                                                                                                                                                                                                                  3
aber s garage inc,(419) 281-5500,44805, , ,                                                                                                                                                                           

In [719]:
data_combined_v3 =data_combined_v2.groupby(['internet_presence']).first().reset_index()

In [720]:
data_combined_v3.shape

(10842, 17)

In [721]:
data_combined_v3.head(5)

internet_presence  \
0  001 putnam chevrolet and cadillac,(479) 856-6390,94031, ,https://maps.google.com/maps?cid=15640886745739450165,                                                                                                       
1  002 fremont chevrolet,(928) 344-2702,94538, ,https://maps.google.com/maps?cid=2859196772727796258,                                                                                                                    
2  003 chevrolet of stevens creek,(520) 625-9414,95117, ,https://maps.google.com/maps?cid=14703233780577702587,                                                                                                          
3  004 capitol chevrolet,(928) 776-1111,95136, ,https://maps.google.com/maps?cid=14703233780577702587,                                                                                                                   
4  1 brookville chevrolet buick,(814) 246-2055,15825, ,https://maps.google.com/maps?cid=12622059353105481409,https://www.facebook.com/brookvillechevy/http://www.cars.com/dealers/154863/1-brookville-chevrolet-buick/   

                                  dealer_addresses  \
0  a sarasota chevrolet new car quotes,bay rd        
1  a sarasota dodge new car quotes,14 holiday dr n   
2  a sarasota honda new car quotes,caribbean dr      
3  a sarasota honda new car quotes,mileground rd     
4  1 brookville chevrolet buick,1 east main st       

                                                             meta_data  \
0  a sarasota chevrolet new car quotes,bay rd,bay rdfl,34239             
1  a sarasota dodge new car quotes,14 holiday dr n,sarasotafl,34231      
2  a sarasota honda new car quotes,caribbean dr,sarasotafl,34231         
3  a sarasota honda new car quotes,mileground rd,mileground rdwv,26505   
4  1 brookville chevrolet buick,1 east main st,brookvillepa,15825        

        id                                 name  tenant_id     address_line  \
0  1525312  a sarasota chevrolet new car quotes  45332      bay rd            
1  1601117  a sarasota dodge new car quotes      83249      14 holiday dr n   
2  1482254  a sarasota honda new car quotes      44849      caribbean dr      
3  1525184  a sarasota honda new car quotes      45332      mileground rd     
4  118193   1 brookville chevrolet buick         2980       1 east main st    

            city state    zip           phone  web  \
0  bay rd         fl    34239  (941) 284-6900  NaN   
1  sarasota       fl    34231  (941) 284-6900  NaN   
2  sarasota       fl    34231  (941) 284-6900  NaN   
3  mileground rd  wv    26505  NaN             NaN   
4  brookville     pa    15825  (814) 849-8313  NaN   

                                              google_url  \
0  https://maps.google.com/maps?cid=15640886745739450165   
1  https://maps.google.com/maps?cid=2859196772727796258    
2  https://maps.google.com/maps?cid=14703233780577702587   
3  https://maps.google.com/maps?cid=14703233780577702587   
4  https://maps.google.com/maps?cid=12622059353105481409   

                                facebook_url  \
0  NaN                                         
1  NaN                                         
2  NaN                                         
3  NaN                                         
4  https://www.facebook.com/brookvillechevy/   

                                                           cars_url zip_short  \
0  NaN                                                               34239      
1  NaN                                                               34231      
2  NaN                                                               34231      
3  NaN                                                               26505      
4  http://www.cars.com/dealers/154863/1-brookville-chevrolet-buick/  15825      

                                                   dealer_contact  
0  a sarasota chevrolet new car quotes,bay rd,(941) 284-6900       
1  a sarasota dodge new car quotes,14 holiday

In [722]:
data_combined_count = data_combined_v3.groupby(['web','google_url','facebook_url','cars_url'])['web','google_url','facebook_url','cars_url'].count()
data_combined_count

web  \
web                            google_url                                            facebook_url                                                         cars_url                                                                         
171nissan.com                  https://maps.google.com/maps?cid=3971356509180237363  https://www.facebook.com/171nissan/                                  https://www.cars.com/dealers/211161/171-nissan/                            1     
academyforddealer.com          https://maps.google.com/maps?cid=11970897632866746619 https://www.facebook.com/academyfordsales                            https://www.cars.com/dealers/2208/academy-ford/                            1     
acrivellisubaru.com            https://maps.google.com/maps?cid=914417773313057702   https://www.facebook.com/acrivelliautogroup/                         https://www.cars.com/dealers/155193/a-crivelli-chevrolet/                  1     
acuraofbayshore.com            https://maps.google.com/maps?cid=40693424694342408    https://www.facebook.com/bayshoreacura/                              https://www.cars.com/dealers/5729/acura-of-bay-shore/                      1     
acuraofberlin.com              https://maps.google.com/maps?cid=5883875942797871769  https://www.facebook.com/acuraberlin/                                https://www.cars.com/dealers/4101/acura-of-berlin/                         1     
...                                                                                                                                                                                                                                 ..     
www.wheelingchevy.com          https://maps.google.com/maps?cid=2923692249652154434  https://www.facebook.com/billstasekchevrolet/                        http://www.cars.com/dealers/602/bill-stasek-chevrolet/                     1     
www.whiteallenaudiofdayton.com https://maps.google.com/maps?cid=279212125703957825   https://www.facebook.com/whitealleneuro/                             https://www.cars.com/dealers/14697/white-allen-european-auto-group/        1     
www.whitemotor.com             https://maps.google.com/maps?cid=10509705568694096173 https://www.facebook.com/bfwhitemotorco/                             http://www.cars.com/dealers/148335/bf-white-motor-company-of-stanton-inc/  1     
www.www.alserra.com            https://maps.google.com/maps?cid=3886489961559017752  https://www.facebook.com/pages/subaru-of-grand-blanc/428560227332988 https://www.cars.com/dealers/5354307/subaru-of-grand-blanc/                1     
wwww.benmynattnissan.com       https://maps.google.com/maps?cid=8381534264661612482  https://www.facebook.com/benmynattnissansalisbury                    https://www.cars.com/dealers/103058/ben-mynatt-nissan/                     1     

                                                                                                                                                                                                                                     google_url  \
web                            google_url                                            facebook_url                                                         cars_url                                                                                
171nissan.com                  https://maps.google.com/maps?cid=3971356509180237363  https://www.facebook.com/171nissan/                                  https://www.cars.com/dealers/211161/171-nissan/                            1            
academyforddealer.com          https://maps.google.com/maps?cid=11970897632866746619 https://www.facebook.com/academyfordsales                            https://www.cars.com/dealers/2208/academy-ford/                            1            
acrivellisubaru.com            https://maps.google.com/maps?cid=914417773313057702   https://www.facebook.com/acrivelliautogroup/                         https://www.cars.com/dealers/15519

In [723]:
data_combined_v3.name.value_counts()

anderson ford                 12
bruckner s mack volvo         10
all american chevrolet        9 
american honda motor co       8 
affordable autos              7 
                             .. 
basin mitsubishi              1 
accessories of florida        1 
all star pontiac gmc truck    1 
avis ford inc legacy          1 
bmw of toledo                 1 
Name: name, Length: 7764, dtype: int64

## Examining why one name is repeated 12 times

In [724]:
data_combined_v3[data_combined_v3.name == 'anderson ford']

internet_presence  \
980  acura of glendale,(818) 502-1100,91204, ,https://maps.google.com/maps?cid=8999307113185777568,https://www.facebook.com/andersonfordclintonilhttps://www.cars.com/dealers/153453/anderson-ford/                                              
981  acura of glendale,(818) 502-1100,91204,www.anderson-ford.net,https://maps.google.com/maps?cid=8999307113185777568,                                                                                                                          
983  acura of glendale,(818) 502-1100,91204,www.andersonfordgibsoncity.com,https://maps.google.com/maps?cid=9078969412338938237,https://www.facebook.com/pages/anderson-ford/126323237423314https://www.cars.com/dealers/199492/anderson-ford/   
984  acura of glendale,(888) 670-9904,91204,www.andersonforddouglas.com,https://maps.google.com/maps?cid=1772629934778698630,https://www.facebook.com/andersonforddouglas/http://www.cars.com/dealers/148152/prince-ford-inc/                    
985  acura of honolulu,(808) 829-3073,96819, ,https://maps.google.com/maps?cid=9078969412338938237,                                                                                                                                              
986  acura of honolulu,(808) 829-3073,96819,www.andersonford.info,https://maps.google.com/maps?cid=18316833063836885106,https://www.facebook.com/anderson-ford-529444370443925https://www.cars.com/dealers/147827/anderson-ford/                 
987  acura of honolulu,(808) 829-3073,96819,www.andersonford.info,https://maps.google.com/maps?cid=5290594378440192604,https://www.facebook.com/anderson-ford-529444370443925                                                                    
988  acura of honolulu,(808) 942-4555,96813, ,https://maps.google.com/maps?cid=1198857466292530781,https://www.facebook.com/andersonfordsc                                                                                                       
989  acura of honolulu,(808) 942-4555,96819,www.andersonford.info,https://maps.google.com/maps?cid=18316833063836885106,                                                                                                                         
990  acura of honolulu,(808) 942-4555,96819,www.andersonfordofsc.com,https://maps.google.com/maps?cid=1198857466292530781,https://www.facebook.com/pages/anderson-ford-mazda/114154978600099https://www.cars.com/dealers/149680/anderson-ford/   
991  acura of huntington,(631) 470-6600,11743, ,https://maps.google.com/maps?cid=10417267975102897205,https://www.facebook.com/justinandersonford/https://www.cars.com/dealers/149680/anderson-ford/                                             
993  acura of huntington,(631) 470-6600,11743,www.andersonford.info,https://maps.google.com/?cid=18316833063836885106,https://www.facebook.com/anderson.ford.77?ref=br_rshttps://www.cars.com/dealers/147827/demontrond-ford/                    

                         dealer_addresses  \
980  anderson ford,1001 state route 10 w    
981  anderson ford,1001 illinois 10         
983  anderson ford,124 s sangamon ave       
984  anderson ford,109 westgreen rd         
985  anderson ford,124 south sangamon ave   
986  anderson ford,13872 us 59              
987  anderson ford,13872 us 59 s            
988  anderson ford,3900 clemson boulevard   
989  anderson ford,13872 u s 59 business    
990  anderson ford,3900 clemson blvd        
991  anderson ford,3900 liberty hwy         
993  anderson ford,3944 u s 59 n bypass     

                                                    meta_data       id  \
980  anderson ford,1001 state route 10 w,clintonil,61727       2825439   
981  anderson ford,1001 illinois 10,clintonil,61727            129655    
983  anderson ford,124 s sangamon ave,gibson cityca,60936      1487454   
984  anderson ford,109 westgreen rd,douglasga,31533            2897231   
985  anderson ford,124 south sangamon ave,gibson cityil,60936  159681    
986  anderson ford,13872 us 59,clevelandmt,7

## There are still mentions of same dealer multiple times, but a combinaton of name, tenant_id and zip might help in reducing the dealer list further, consolidating them

In [725]:
data_combined_v4 =data_combined_v3.groupby(['name','city','zip_short']).first().reset_index()

In [726]:
data_combined_v4.shape

(8342, 17)

### Data further reduced to 8342

In [727]:
data_combined_v4.isnull().sum()

name                 0   
city                 0   
zip_short            0   
internet_presence    0   
dealer_addresses     0   
meta_data            0   
id                   0   
tenant_id            0   
address_line         0   
state                0   
zip                  0   
phone                900 
web                  2038
google_url           1127
facebook_url         2723
cars_url             2986
dealer_contact       0   
dtype: int64

In [728]:
data_combined_v4[data_combined_v4.name == 'anderson ford']

name         city zip_short  \
1369  anderson ford  anderson     29621      
1370  anderson ford  cleveland    77327      
1371  anderson ford  cleveland    77328      
1372  anderson ford  clinton      61727      
1373  anderson ford  douglas      31533      
1374  anderson ford  gibson city  60936      

                                                                                                                                                                                                                              internet_presence  \
1369  acura of honolulu,(808) 942-4555,96813, ,https://maps.google.com/maps?cid=1198857466292530781,https://www.facebook.com/andersonfordsc                                                                                                       
1370  acura of honolulu,(808) 829-3073,96819,www.andersonford.info,https://maps.google.com/maps?cid=18316833063836885106,https://www.facebook.com/anderson-ford-529444370443925https://www.cars.com/dealers/147827/anderson-ford/                 
1371  acura of honolulu,(808) 829-3073,96819,www.andersonford.info,https://maps.google.com/maps?cid=5290594378440192604,https://www.facebook.com/anderson-ford-529444370443925                                                                    
1372  acura of glendale,(818) 502-1100,91204, ,https://maps.google.com/maps?cid=8999307113185777568,https://www.facebook.com/andersonfordclintonilhttps://www.cars.com/dealers/153453/anderson-ford/                                              
1373  acura of glendale,(888) 670-9904,91204,www.andersonforddouglas.com,https://maps.google.com/maps?cid=1772629934778698630,https://www.facebook.com/andersonforddouglas/http://www.cars.com/dealers/148152/prince-ford-inc/                    
1374  acura of glendale,(818) 502-1100,91204,www.andersonfordgibsoncity.com,https://maps.google.com/maps?cid=9078969412338938237,https://www.facebook.com/pages/anderson-ford/126323237423314https://www.cars.com/dealers/199492/anderson-ford/   

                          dealer_addresses  \
1369  anderson ford,3900 clemson boulevard   
1370  anderson ford,13872 us 59              
1371  anderson ford,13872 us 59 s            
1372  anderson ford,1001 state route 10 w    
1373  anderson ford,109 westgreen rd         
1374  anderson ford,124 s sangamon ave       

                                                  meta_data       id  \
1369  anderson ford,3900 clemson boulevard,andersonsc,29621  2821208   
1370  anderson ford,13872 us 59,clevelandmt,77327            1491324   
1371  anderson ford,13872 us 59 s,clevelandtx,77328          3238506   
1372  anderson ford,1001 state route 10 w,clintonil,61727    2825439   
1373  anderson ford,109 westgreen rd,douglasga,31533         2897231   
1374  anderson ford,124 s sangamon ave,gibson cityca,60936   1487454   

      tenant_id            address_line state    zip           phone  \
1369  119401     3900 clemson boulevard  sc    29621  (864) 810-1036   
1370  44849      13872 us 59             mt    77327  (888) 871-9716   
1371  130469     13872 us 59 s           tx    77328  (281) 545-7162   
1372  119451     1001 state route 10 w   il    61727  (877) 883-1717   
1373  289        109 westgreen rd        ga    31533  (912) 384-2600   
1374  44849      124 s sangamon ave      ca    60936  (217) 784-4119   

                                 web  \
1369  www.andersonfordofsc.com         
1370  www.andersonford.info            
1371  www.andersonford.info            
1372  www.anderson-ford.net            
1373  www.andersonforddouglas.com      
1374  www.andersonfordgibsoncity.com   

                                                 google_url  \
1369  https://maps.google.com/maps?cid=1198857466292530781    
1370  https://maps.google.com/maps?cid=18316833063836885106   
1371  https://maps.google.com/maps?cid=5290594378440192604    
1372  https://maps.google.com/maps?cid=8999307113185777568    
1373  https://maps.google.com/maps?cid=1772629934778698630   

In [729]:
data_combined_v4.name.value_counts()

bruckner s mack volvo                  10
american honda motor co                8 
affordable autos                       7 
affordable auto sales                  7 
auto smart                             6 
                                      .. 
bergstrom mercedes benz of appleton    1 
audi stevens creek                     1 
bmw of santa maria                     1 
bison motor company                    1 
audi las vegas                         1 
Name: name, Length: 7764, dtype: int64

In [730]:
data_combined_v4[data_combined_v4.name == 'bruckner s mack volvo']

name          city zip_short  \
8089  bruckner s mack volvo  albuquerque   87121      
8090  bruckner s mack volvo  amarillo      79118      
8091  bruckner s mack volvo  bossier city  71111      
8092  bruckner s mack volvo  farmington    87401      
8093  bruckner s mack volvo  garden city   67846      
8094  bruckner s mack volvo  hays          67601      
8095  bruckner s mack volvo  hobbs         88240      
8096  bruckner s mack volvo  lubbock       79404      
8097  bruckner s mack volvo  monroe        71203      
8098  bruckner s mack volvo  odessa        79766      

                                                                                                                                                                                        internet_presence  \
8089  autonation ford lincoln wolfchase 2363, ,38133,www.brucknertruck.com,https://maps.google.com/maps?cid=7718685067242739490,                                                                            
8090  autonation ford lincoln wolfchase 2363, ,38133,www.brucknertruck.com,https://maps.google.com/maps?cid=5404558478043464866,https://www.facebook.com/brucknertrucksales/                                
8091  autonation ford lincoln wolfchase 2363, ,38133,www.brucknertruck.com,https://maps.google.com/maps?cid=11146096563401691841,https://www.facebook.com/pages/bruckner-truck-sales-inc/245550159595462    
8092  autonation ford lincoln wolfchase 2363, ,38133,www.brucknertruck.com,https://maps.google.com/maps?cid=9380787747266475927,https://www.facebook.com/brucknertruckfarmingtonnm/                         
8093  autonation ford lincoln wolfchase,(901) 209-1257,38133,www.brucknertruck.com,https://maps.google.com/maps?cid=10582764957551610916,                                                                   
8094  autonation ford lincoln union city 2325, ,30291,www.brucknertruck.com,https://maps.google.com/maps?cid=8959750749014169057,https://www.facebook.com/brucknerhays/                                     
8095  autonation ford lincoln union city 2325, ,30291,www.brucknertruck.com,https://maps.google.com/maps?cid=11187758810578127375,https://www.facebook.com/pages/bruckners-truck-sales/167522453260378      
8096  autonation ford lincoln union city 2325, ,30291,www.brucknertruck.com,https://maps.google.com/maps?cid=13516605791043902826,                                                                          
8097  autonation ford lincoln wolfchase,(901) 382-5555,38133,www.brucknertruck.com,https://maps.google.com/maps?cid=9658066730056565244,                                                                    
8098  autonation ford lincoln union city 2325, ,30291,www.brucknertruck.com,https://maps.google.com/maps?cid=17086503579344920412,                                                                          

                                  dealer_addresses  \
8089  bruckner s mack volvo,8101 daytona rd nw       
8090  bruckner s mack volvo,9471 i 40 east exit 76   
8091  bruckner s mack volvo,4275 meadow ln           
8092  bruckner s mack volvo,4050 bloomfield hwy      
8093  bruckner s mack volvo,2509 n taylor ave        
8094  bruckner s mack volvo,2101 commerce pkwy       
8095  bruckner s mack volvo,2121 nw county rd        
8096  bruckner s mack volvo,1125 slaton rd           
8097  bruckner s mack volvo,2602 millhaven rd        
8098  bruckner s mack volvo,2441 e interstate 20     

                                                          meta_data       id  \
8089  bruckner s mack volvo,8101 daytona rd nw,albuquerquenm,87121   1594486   
8090  bruckner s mack volvo,9471 i 40 east exit 76,amarillotx,79118  1518181   
8091  bruckner s mack volvo,4275 meadow ln,bossier cityla,71111      1507833   
8092  bruckner s mack volvo,4050 bloomfield hwy,farmingtonnm,87401   2739739   
8093  bruckner s mack volvo,2509 n taylor ave,garden cityks,67846    2743175   
8094  bruckner s mack volvo,2101 commerce pkwy,haysks,67601          1513651   
8095  

In [731]:
data_combined_v4.shape

(8342, 17)

### Lets try to consolidate based on web address to further refine

In [733]:
data_combined_v4.web.head(10)

0    www.bigotires.com           
1    NaN                         
2    NaN                         
3    NaN                         
4    www.brookvillegmdealer.com  
5    www.brookvillegmdealer.com  
6    NaN                         
7    NaN                         
8    www.cochranofmonroeville.com
9    www.cochranofrobinson.com   
Name: web, dtype: object

In [734]:
data_combined_v4.web.value_counts()

www.billionauto.com                        20
www.brucknertruck.com                      11
www.autonationcollisioncenters.com         11
www.autonation.com                         7 
www.honda.com                              6 
                                          .. 
www.actionautochryslerdodgejeep.com        1 
www.bobmooreaudiok.com                     1 
www.billybenderchryslerjeepdodgeram.com    1 
www.billingsleyflm.com                     1 
www.audimv.com                             1 
Name: web, Length: 4215, dtype: int64

In [735]:
data_combined_v4[data_combined_v3.web == 'www.billionauto.com']

name              city zip_short  \
7263  bosak chrysler                        merrillville      46410      
7264  bosak ford                            chesterton        46304      
7265  bosak honda                           highland          46322      
7266  bosak honda                           michigan city     46360      
7267  bosak honda highland                  highland          46322      
7273  bosak nissan                          burns harbor      46304      
7274  boshears ford                         marshall          49068      
7275  boshears ford sales                   marshall          49068      
7276  boshears ford sales 09756             marshall          49068      
7277  boshears ford sales inc               marshall          49068      
7278  boss chrysler plymouth jeep 11364     long island city  11101      
7279  boss motors                           marysville        66508      
7280  boss motors 05182                     marysville        66508      
7286  boston volkswagen                     watertown         2472       
7312  boucher nissan waukesha               waukesha          53186      
7324  boulder toyota                        boulder           80301      
7325  boulevard auto sales inc              georgetown        19947      
7326  boulevard auto sales inc parts        georgetown        19947      
7333  boulevard ford lincoln 01268          georgetown        19947      
7334  boulevard ford lincoln of georgetown  georgetown        19947      
7335  boulevard ford lincoln of lewes       lewes             19958      
7336  boulevard lincoln                     georgetown        19947      
7341  bournival inc                         portsmouth        3801       

                                                                                                                                                                                                                                                      internet_presence  \
7263  autonation acura spokane valley service center,(509) 289-5612,99212, ,https://maps.google.com/?cid=9588476791745658, http://www.cars.com/dealers/155307/bosak-chrysler-jeep-dodge-ram/                                                                              
7264  autonation acura stevens creek,(408) 758-5502,95051,www.lakeshoreford.net,https://maps.google.com/maps?cid=10056049863143110313,https://www.facebook.com/179118028045https://www.cars.com/dealers/14973/lake-shore-ford/                                            
7265  autonation acura stevens creek service center,(408) 780-1652,95051,www.bosakhondahighland.com,https://maps.google.com/maps?cid=16265415469065877319,https://www.facebook.com/167889936602685https://www.cars.com/dealers/82/bosak-honda/                            
7266  autonation acura stevens creek 2674, ,95051,www.bosakhondamc.com,https://maps.google.com/maps?cid=7865471118760769803,https://www.facebook.com/314891999293https://www.cars.com/dealers/15007/bosak-honda-in-michigan-city/                                         
7267  autonation acura stevens creek,(408) 758-5502,95051,www.bosakhondahighland.com,https://maps.google.com/maps?cid=16265415469065877319,https://www.facebook.com/bosakhondahighland/https://www.cars.com/dealers/82/bosak-honda/                                       
7273  autonation acura of stevens creek,(408) 758-5502,95051,www.bosaknissan.com,https://maps.google.com/maps?cid=14062767460558294530,https://www.facebook.com/chestertonnissan/https://www.cars.com/dealers/198477/bob-rohrman-nissan-of-chesterton-in-burns-harbor/    
7274  autonation alfa romeo north denver parts center,(844) 427-4422,80234,www.boshearsford.com,https://maps.google.com/maps?cid=15208119511261563040,https://www.facebook.com/boshears-ford-368609539905151https://www.cars.com/dealers/16020/boshears-ford-sales-inc/   
7275  autonation acura of stevens creek,(408) 758-5502,95051,www.boshearsford.com,https://maps.google.com/maps?cid=1520811951

## No notable same references found on manual vision but still...

###  Slowly it can be refined further with more time..  with given time constraints, I stop here. Now lets build a system to map the canonical list to new entries.



# Lets assign an ID for the canonical list

In [507]:
data_combined_v4['canonical_id'] = data_combined_v3.index + 1

In [509]:
data_combined_v3.columns

Index(['name', 'city', 'zip_short', 'dealer_addresses', 'meta_data', 'id',
       'tenant_id', 'address_line', 'state', 'zip', 'phone', 'web',
       'google_url', 'facebook_url', 'cars_url', 'dealer_contact',
       'internet_presence', 'canonical_id'],
      dtype='object')

In [512]:
canonical_df =data_combined_v3[['canonical_id','name','city','zip_short','address_line','state','zip','phone','web','google_url','facebook_url','cars_url']]

##  The process has following methods.

##### Clean and refine data.
##### compare it with canonical list.
##### Get top 1 value with its confidence.
#### If confidence is well below certain level, we should update and add it to canonical list.

# The entire functional logic is written in datamapper.py



## Building a model for comparision can be done through certain similarty techniques like : 

1. Levenstein distance
2. Cosine Similarity
3. Fuzzy search etc.


Lets build confidence based on similarity score.

Implementing FuzzyWuzzy for a test data as below

In [736]:
canonical_df.columns

Index(['canonical_id', 'name', 'city', 'zip_short', 'address_line', 'state',
       'zip', 'phone', 'web', 'google_url', 'facebook_url', 'cars_url',
       'meta_data'],
      dtype='object')

In [513]:
canonical_df['meta_data'] = canonical_df['name'].fillna(' ') + ',' + canonical_df['address_line'].fillna(' ') + ',' + canonical_df['city'].fillna(' ')\
+canonical_df['state'].fillna(' ') + ',' + canonical_df['zip_short'].fillna(' ') + ',' + canonical_df['web'].fillna(' ') + ',' + canonical_df['phone'].fillna(' ')

C:\Users\ynagubandi\AppData\Roaming\Python\Python36\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [514]:
canonical_df['meta_data'].head(5)

0    001 putnam chevrolet and cadillac,3052 martin ...
1    002 fremont chevrolet,5850 cushing pkwy,fremon...
2    003 chevrolet of stevens creek,3640 stevens cr...
3    004 capitol chevrolet,905 capitol expressway a...
4    1 brookville chevrolet,1 e main st,brookvillep...
Name: meta_data, dtype: object

In [518]:
test = data['meta_data'].iloc[1]
test

'1 brookville chevrolet buick,1 e main st,brookvillepa,15825'

In [532]:
fuzz_list = canonical_df['meta_data'].apply(lambda x: fuzz.ratio(x,test))

In [560]:
max(fuzz_list)

81

In [531]:
fuzz_list = max(fuzz_list)

In [562]:
print(canonical_df[canonical_df.canonical_id == (fuzz_list[fuzz_list == max(fuzz_list)].index[0]+1)][['meta_data','canonical_id']])

                                                                                meta_data  \
5  1 brookville chevrolet buick,1 e main st,brookvillepa,15825,www.brookvillegmdealer.com   

   canonical_id  
5  6             


In [558]:
test

'1 brookville chevrolet buick,1 e main st,brookvillepa,15825'

In [740]:
process.extract(test,canonical_df.meta_data,scorer=fuzz.token_sort_ratio)

In [745]:
test_answer

('1 brookville chevrolet buick,1 e main st,brookvillepa,15825,www.brookvillegmdealer.com',
 81,
 5)

## Both are reasonablly similar, so started implementing Data mapper at this point.

# Test the class and functions created Datamapper.py .


In [17]:
import datamapper

In [18]:
sample = data[0:1000]

In [12]:
canonical_list = datamapper.prepare_canonical_dataframe(sample)

Begin data cleaning
The id has following unique elements :979 out of 30000 entries
The name has following unique elements :321 out of 30000 entries
The tenant_id has following unique elements :87 out of 30000 entries
The address_line has following unique elements :377 out of 30000 entries
The city has following unique elements :210 out of 30000 entries
The state has following unique elements :44 out of 30000 entries
The zip has following unique elements :260 out of 30000 entries
The phone has following unique elements :291 out of 30000 entries
The web has following unique elements :263 out of 30000 entries
The google_url has following unique elements :271 out of 30000 entries
The facebook_url has following unique elements :300 out of 30000 entries
The cars_url has following unique elements :273 out of 30000 entries
data cleaned
data reduced to 320


# For first 1000 entries canonical list has been reduced to 320.

Now lets map them back with the 100 entries

In [19]:
map_raw_data = datamapper.map_data_to_canonical_data(sample, canonical_list)

Begin data cleaning
The id has following unique elements :979 out of 30000 entries
The name has following unique elements :321 out of 30000 entries
The tenant_id has following unique elements :87 out of 30000 entries
The address_line has following unique elements :377 out of 30000 entries
The city has following unique elements :210 out of 30000 entries
The state has following unique elements :44 out of 30000 entries
The zip has following unique elements :260 out of 30000 entries
The phone has following unique elements :291 out of 30000 entries
The web has following unique elements :263 out of 30000 entries
The google_url has following unique elements :271 out of 30000 entries
The facebook_url has following unique elements :300 out of 30000 entries
The cars_url has following unique elements :273 out of 30000 entries
data cleaned
(1000, 2)


In [22]:
map_raw_data[['meta_data','mapping_canonical_id','confidence']]

meta_data  \
0    1 brookville chevrolet,1 e main st,brookvillepa,15825,www.brookvillegmdealer.com,(814) 849-8313           
1    1 brookville chevrolet buick,1 e main st,brookvillepa,15825, ,(814) 849-8313                              
2    1 brookville chevrolet buick,1 east main st,brookvillepa,15825, ,(814) 849-8313                           
3    1 brookville chevrolet buick,1 e main st,brookvillepa,15825, ,(814) 849-8313                              
4    1 brookville chevrolet buick,30 main st,brookvillepa,15825,www.brookvillegmdealer.com,(814) 849-8313      
..                                                                                                    ...      
995  armen chevrolet saab of ardmore,125 lancaster ave,ardmorepa,19003, ,(610) 649-0300                        
996  arnell chevrolet inc,239 melton rd,chestertonin,46304, ,(219) 787-9300                                    
997  arnell chevrolet inc,239 melton rd,chestertonin,46304, ,(219) 787-9300                                    
998  arnell chevrolet inc,239 melton rd,burns harborin,46304, ,(219) 787-9300                                  
999  arnie bauer buick gmc cadillac,5525 miller circle dr,mattesonil,60443,www.arniebauer.com,(708) 720-1900   

    mapping_canonical_id confidence  
0    5                    100        
1    6                    84         
2    6                    83         
3    6                    84         
4    6                    96         
..  ..                    ..         
995  318                  96         
996  319                  82         
997  319                  82         
998  319                  100        
999  320                  100        

[1000 rows x 3 columns]

In [25]:
map_raw_data['data_quality'] = 100-(map_raw_data.isnull().sum(axis=1)/len(map_raw_data.columns))*100

In [26]:
map_raw_data['data_quality']

0      100.00
1      81.25 
2      93.75 
3      75.00 
4      100.00
        ...  
995    93.75 
996    75.00 
997    81.25 
998    81.25 
999    100.00
Name: data_quality, Length: 1000, dtype: float64

# Conclusion :

1. Data cleaned and refined

2. Best estimate of cannical list extracted from data.

3. All other metrics like web address,google_urls and websites are well preserved wherever available through data consolidation.

4. System created with a new primary key canonical_id and a model based on similarty score for creating mappings 

5. A new script datamapper.py is created to map the raw_data based on canonical list generated, which can be joined using canonical_id in list and mapping_canonical_id in raw_data.

6. Data quality computed for raw_data

7. Provision to update canonical list is also provided to scale the service when in production.

8. 3 suggestive deployment ways are created (Microservice approach, Heroku based Deployment, Lambda in AWS)

9. Further scope and direction provided in word document.

10. Operational instructions for datamapper and how it can be further improved provided in walkthrough document



# Playground (for rough work)

In [496]:
import re

In [497]:
sentance = "hello there 1 is here for you"
re.sub(r'(?<!\S)(\d)(?!\S)' , r'_suff.\2',sentance)

error: invalid group reference 2 at position 7

In [ ]:
# ROUGH

# pattern = r'\b({})\b'.format('|'.join(sorted(re.escape(number) for number in number_map)))
# address2 = "one ,123 north anywhere southstreet"
# re.sub(pattern, lambda m: number_map.get(m.group(0).lower()), address2, flags=re.IGNORECASE)


In [ ]:
#df['date'].replace('^([0-9])$', number_map'\1', regex=True)
#data['meta_data'] = data['meta_data'].replace('(\s|,)?(\d)(\s|,)')
#data['meta_data'].astype(str).apply(lambda row: re.sub(r'(?<!\S)\d+(?!\S)', lambda x: p.number_to_words(x.group()), row))